### ISSUE TITLE
add a `LazyProvingKey` and rework loading of proving keys

### ISSUE THREAD

**Issue by redshiftzero** [penumbra-zone/penumbra#2712]
Currently the proving keys are only present in the `penumbra-proof-params` crate if the `proving-keys` feature is enabled: that feature uses the [`include_bytes!` macro](https://doc.rust-lang.org/std/macro.include_bytes.html) to bundle the generated proving key files.  The proving keys are each of the type `Lazy<ProvingKey<..>>`.

Instead we could always have the proving keys be present, i.e. `SPEND_PROOF_PROVING_KEY`, etc. will no longer be conditionally included if the `proving-keys` feature is enabled. The `proving-keys` feature would instead control whether the proving key object is loaded in the crate.

We'd create a `LazyProvingKey` type that internally holds a `OnceCell<ProvingKey>`. It behaves just like the `Lazy` type: it implements `Deref`, except if you dereference it without the `proving-keys` feature enabled or without loading the proving key from elsewhere, it will panic.

We’d have an API, e.g. a method on `LazyProvingKey::load` that takes a byte slice. If the `proving-keys` feature was not enabled, you must call this loading method. Internally it will call `get_or_init` on the `OnceCell`: https://docs.rs/once_cell/latest/once_cell/sync/struct.OnceCell.html#method.get_or_init.

Tl;dr: if you want everything bundled together, you enable the `proving-keys` feature as before, and everything works in exactly the same way. If you _don’t_ use that feature, then it's your responsibility to call the load method with the bytes of the proving key before doing anything with the proving keys, else the program panics. The benefit of this is that you have the option of loading from a file, and not having to balloon the size of the binary.

**Comment by conorsch** [comment]
Some context: the dapp website was previously loading a ~90MB wasm file, which delayed TTFP. That's no longer the case, given that Zpoken retooled things so that the wasm logic is consumed exclusively by the extension, rather than by the dapp. It'd still be nice to resolve this problem, but let's wait until we have some dedicated web story support.

**Comment by cronokirby** [comment]
Another additional thing we could do here is always encode the hash of the proving keys in the binary itself, that way the binary acts as a root of trust for this information. This way, the loading method can be trustless, like via a public API we control, or any other kind of CDN. The API proposed above would also work, just that the `load` method would also check that the hash of the bytes passed in matches.

**Comment by hdevalence** [comment]
I think we do currently encode the hashes of the proving and verifying keys, IIRC it's how our auto-download-during-build mechanism works (I believe we check that the data we fetch from the network matches the expected hashes before we embed it into the binary).

**Comment by cronokirby** [comment]
My understanding, from reading https://github.com/penumbra-zone/penumbra/blob/main/crates/crypto/proof-params/build.rs, is that we have a `X.bin` file for each proof which stores a Git LFS pointer to actual proving key, and then we can manually resolve those at build time by parsing the pointer and doing the download ourselves. Now, this pointer does encode a SHA-256, so yeah, we are committing to the proving keys, at least.

We also don't support lazy verification keys, so these don't have any hashes either, I think.

**Comment by hdevalence** [comment]
It's not just the LFS pointer, but also we encode hashes of the keys directly in the source code, eg https://rustdoc.penumbra.zone/main/penumbra_proof_params/spend/constant.PROVING_KEY_ID.html

**Comment by cronokirby** [comment]
Ah, I was missing that part!

**Comment by redshiftzero** [comment]
Closing (done in #3337)
